File: TypeScript/Practice/LeetCode/Medium/Find Players With Zero or One Losses.ts
// https://leetcode.com/problems/find-players-with-zero-or-one-losses/description/?envType=daily-question&envId=2024-01-15

function findWinners(matches: number[][]): number[][] {
    const playersOne: number[] = [];
    const playersZero: number[] = [];
    const playersLoses: {[index: string] : number} = {};

    for (let i = 0; i < matches.length; i++) {
        if (playersLoses[matches[i][1]] === undefined) {
            playersLoses[matches[i][1]] = 1;
        } else {
            playersLoses[matches[i][1]] += 1;
        }
    }

    for (let player in playersLoses) {
        if (playersLoses[player] === 1) {
            playersOne.push(Number(player));
        }
    }

    for (let i = 0; i < matches.length; i++) {
        if (playersLoses[matches[i][0]] === undefined && !playersZero.includes(matches[i][0])) {
            playersZero.push(matches[i][0]);
        }
    }

    playersOne.sort((a, b) => a - b);
    playersZero.sort((a, b) => a - b);

    return [[...playersZero], [...playersOne]];    
};

File: C#/Practice/LeetCode/Medium/Find Players With Zero or One Losses.cs
// https://leetcode.com/problems/find-players-with-zero-or-one-losses/description/?envType=daily-question&envId=2024-01-15

public class Solution {
    public IList<IList<int>> FindWinners(int[][] matches) {
        List<int> playersOne = new List<int>();
        List<int> playersZero = new List<int>();
        Dictionary<int, int> playersLoses = new Dictionary<int, int>();

        foreach (var match in matches)
        {
            if (!playersLoses.ContainsKey(match[1]))
            {
                playersLoses[match[1]] = 1;
            }
            else
            {
                playersLoses[match[1]] += 1;
            }
        }

        foreach (var player in playersLoses.Keys)
        {
            if (playersLoses[player] == 1)
            {
                playersOne.Add(player);
            }
        }

        foreach (var match in matches)
        {
            if (!playersLoses.ContainsKey(match[0]) && !playersZero.Contains(match[0]))
            {
                playersZero.Add(match[0]);
            }
        }

        playersOne.Sort();
        playersZero.Sort();

        return new int[][] { playersZero.ToArray(), playersOne.ToArray() };   
    }
}